h2o-genmodel/src/main/java/hex/genmodel/easy/RowToRawDataConverter.java

```diff
@@ -66,110 +51,67 @@ public RowToRawDataConverter(GenModel m,
-            try {
-              img = isURL? ImageIO.read(new URL(s)) : ImageIO.read(new File(s));
-            }
-            catch (IOException e) {
-              throw new PredictException("Couldn't read image from " + s);
-            }
-          } else if (isText) {
-            // TODO: use model-specific vectorization of text
-            throw new PredictException("MOJO scoring for text classification is not yet implemented.");
-          }
-          else {
-            // numeric
-            try {
-              value = Double.parseDouble(s);
-            } catch(NumberFormatException nfe) {
-              if (!_convertInvalidNumbersToNa)
-                throw new PredictNumberFormatException("Unable to parse value: " + s + ", from column: "+ dataColumnName + ", as Double; " + nfe.getMessage());
-            }
-          }
-        } else if (o instanceof Double) {
-          value = (Double) o;
-        } else if (o instanceof byte[] && isImage) {
-          // Read the image from raw bytes
-          InputStream is = new ByteArrayInputStream((byte[]) o);
-          try {
-            img = ImageIO.read(is);
-          } catch (IOException e) {
-            throw new PredictException("Couldn't interpret raw bytes as an image.");
-          }
-        } else {
-          throw new PredictUnknownTypeException(
-                  "Unexpected object type " + o.getClass().getName() + " for numeric column " + dataColumnName);
-        }
-
-        if (isImage && img != null) {
-          DeepwaterMojoModel dwm = (DeepwaterMojoModel) _m;
-          int W = dwm._width;
-          int H = dwm._height;
-          int C = dwm._channels;
-          float[] _destData = new float[W * H * C];
-          try {
-            GenModel.img2pixels(img, W, H, C, _destData, 0, dwm._meanImageData);
-          } catch (IOException e) {
-            e.printStackTrace();
-            throw new PredictException("Couldn't vectorize image.");
-          }
-          rawData = new double[_destData.length];
-          for (int i = 0; i < rawData.length; ++i)
-            rawData[i] = _destData[i];
-          return rawData;
-        }
+      Object o = data.get(dataColumnName);
+      if (convertValue(dataColumnName, o, _domainMap.get(index), index, rawData)) {
+        return rawData;
+      }
+    }
+    return rawData;
+  }
 
-        if (Double.isNaN(value)) {
-          // If this point is reached, the original value remains NaN.
-          _errorConsumer.dataTransformError(dataColumnName, o, "Given non-categorical value is unparseable, treating as NaN.");
+  protected boolean convertValue(String columnName, Object o, CategoricalEncoder catEncoder,
+                                 int targetIndex, double[] rawData) throws PredictException {
+    if (catEncoder == null) {
+      // Column is either numeric or a string (for images or text)
+      double value = Double.NaN;
```

Comment:
Value could be guaranteed to be assigned just once, making it effectively final. Can we assign NaN only in the else block ?

Reply:
This is opinion based. Minor stuff, don't worry about it too much please.

Reply:
I tried that but it made the code actually worse - because of the try-catch where we are handling the NumberFormatException

will keep as-is